tensorflow/python/keras/layers/recurrent_v2.py

```diff
@@ -486,9 +486,12 @@ def step(cell_inputs, cell_states):
 def cudnn_gru(inputs, init_h, kernel, recurrent_kernel, bias, mask, time_major,
               go_backwards):
   """GRU with CuDNN implementation which is only available for GPU."""
-  if not time_major:
+  if not time_major and mask is None:
     inputs = array_ops.transpose(inputs, perm=(1, 0, 2))
```

Comment:
This PR will skip the transpose when sequence_lengths/mask is provided (indicating the v3 API will be used and only v3 supports batch major).

Reply:
Agreed, I wasn't notice this when I look at the start of the function. I can approve the PR as is, but I think it will be nice to have v3 enabled for non-masking case as well.

Reply:
Sure. Actually, the cudnn rnn ops in contrib has already done it by preparing the seq_len for users. https://github.com/tensorflow/tensorflow/blob/a232fda68d692da0f409608dd5960d63ac700e5d/tensorflow/contrib/cudnn_rnn/python/ops/cudnn_rnn_ops.py#L1133.

I can do another PR for this feature when this is done.

Reply:
oh, I was wondering if this can be done on the c op level for performance reason. It will be nice If the implementation of cudnnrnn_v3 can take care of the absent of sequence length rather than force user to provide one.

Reply:
I think it could be done. But I don't see the performance issue for letting python prepare the seq_len, since this array is on host memory and will be on host memory even for cudnn calls.

Reply:
I guess you have better knowledge for the cudnn kernel, I original idea was that for non-mask input, the generation of seq_len tensor is not needed at all (just process the whole input sequence).

Reply:
Yes, let me check how much work need to be done for generating seq_len inside the op (which might require the API change, because V3 requires seq_len).

But for this PR, I think it is ready to go.